airflow/www_rbac/security.py

```diff
@@ -120,60 +138,266 @@
+
+        user_perms_views = self.get_all_permissions_views()
+        # return all dags that the user could access
+        return set([view for perm, view in user_perms_views if perm in dag_perms])
+
+    def has_access(self, permission, view_name, user=None):
+        """
+        Verify whether a given user could perform certain permission
+        (e.g can_read, can_write) on the given dag_id.
+
+        :param str permission: permission on dag_id(e.g can_read, can_edit).
+        :param str view_name: name of view-menu(e.g dag id is a view-menu as well).
+        :param str user: user name
+        :return: a bool whether user could perform certain permission on the dag_id.
+        """
+        if not user:
+            user = g.user
+        if user.is_anonymous():
+            return self.is_item_public(permission, view_name)
+        return self._has_view_access(user, permission, view_name)
+
+    def clean_perms(self):
+        """
+        FAB leaves faulty permissions that need to be cleaned up
+        """
+        logging.info('Cleaning faulty perms')
+        sesh = self.get_session
+        pvms = (
+            sesh.query(sqla_models.PermissionView)
+            .filter(or_(
+                sqla_models.PermissionView.permission == None,  # NOQA
+                sqla_models.PermissionView.view_menu == None,  # NOQA
+            ))
+        )
+        deleted_count = pvms.delete()
+        sesh.commit()
+        if deleted_count:
+            logging.info('Deleted {} faulty permissions'.format(deleted_count))
+
+    def _merge_perm(self, permission_name, view_menu_name):
+        """
+        Add the new permission , view_menu to ab_permission_view_role if not exists.
+        It will add the related entry to ab_permission
+        and ab_view_menu two meta tables as well.
+
+        :param str permission_name: Name of the permission.
+        :param str view_menu_name: Name of the view-menu
+
+        :return:
+        """
+        permission = self.find_permission(permission_name)
+        view_menu = self.find_view_menu(view_menu_name)
+        pv = None
+        if permission and view_menu:
+            pv = self.get_session.query(self.permissionview_model).filter_by(
+                permission=permission, view_menu=view_menu).first()
+        if not pv and permission_name and view_menu_name:
+            self.add_permission_view_menu(permission_name, view_menu_name)
+
+    def create_custom_dag_permission_view(self):
+        """
+        Workflow:
+        1. when scheduler found a new dag, we will create an entry in ab_view_menu
+        2. we fetch all the roles associated with dag users.
+        3. we join and create all the entries for ab_permission_view_menu
+           (predefined permissions * dag-view_menus)
+        4. Create all the missing role-permission-views for the ab_role_permission_views
+
+        :return: None.
+        """
+        # todo(Tao): should we put this function here or in scheduler loop?
```

Comment:
I think Max @mistercrunch 's concern is that if we want to let scheduler calls this method, we need to bring FAB models into airflow core model file, otherwise everytime scheduler(which could run on a different machine) needs to instantiate securityManager which needs to create/instantiate the Flask-FAB web which is not what we want. I think to simplify the use case, we could adopt what Max first proposed approach which all the acl handles live inside the webserver, and then later re-write cli to call webserver api endpoint for DAG acl access. Joy @jgao54 , What do you think?

Reply:
Thanks for getting me on the same page. Makes sense now, I'm okay with having all acl handles live inside webserver for now to unblock this PR.

Reply:
Yeah for now I'm thinking that RBAC is strictly a web construct. That has implications like:
* CLI would get RBAC features by moving to using REST
* scheduler is unaware of roles & users, maybe as the supervisor it periodically calls a REST endpoint to trigger creating new perms from new DAGs
* Currently any RBAC operation requires having the flask `app` object created, imports a whole lot of stuff and has an expensive init cycle, we clearly don't want to trigger that in any other way than starting up the web server. So if we wanted to access RBAC / perms from outside the web server, we'd have to stub the roles/user/perms ORM models in models.py, I'd like to try an avoid doing this by making the strict rule that RBAC is a web construct

Reply:
So in short. Maybe we can add a `sync_perms` endpoint that takes care of that and maybe the scheduler calls it periodically.

Reply:
I like the idea of making RBAC a web construct for now and creating a sync_perm endpoint. I think this means we'd need to create a "System" role (superior to the current admin role) in order to prevent any user from having access to it, which I think is reasonable.

Reply:
thanks @mistercrunch , I will create a todo for the sync_perms endpoint.